pgjdbc/src/test/java/org/postgresql/test/jdbc2/optional/BaseDataSourceTest.java

```diff
@@ -218,6 +219,32 @@ public void testJndi() {
         oldurl, url);
   }
 
+  /**
+   * tests that failover urls survive the parse/rebuild roundtrip with and without specific ports
+   */
+  @Test
+  public void testFailoverUrl() throws NamingException, ClassNotFoundException, IOException {
```

Comment:
This test is mutating the member variable `bds`. Depending on the order the tests are run it could cause other tests to fail. It should be creating it's own local instance and not depend on or mutate anything outside the test method.

So something like this:

```java
private static testFailoverUrl(String url, String expectedUrl) {
    BaseDataSource ds = new BaseDataSource();
    ds.setUrl(url);
    assertEquals("BaseDataSource.getURL() generated via setUrl(...) should match", expectedUrl, ds.getURL().replaceAll("\\?.*$", ""));
    
    BaseDataSource fromRefDs = new BaseDataSource();
    fromRefDs.setFromReference(ds);
    assertEquals("BaseDataSource.getURL() generated setFromReference(...) should match", expectedUrl, fromRefDs.getURL().replaceAll("\\?.*$", ""));

    BaseDataSource initializeFromDs = new BaseDataSource();
    initializeFromDs.initializeFrom(ds);
    assertEquals("BaseDataSource.getURL() generated initializeFrom(...) should match", expectedUrl, fromRefDs.getURL().replaceAll("\\?.*$", ""));
}

@Test
public void testFailoverUrl() throws NamingException, ClassNotFoundException, IOException {
    testFailoverUrl("jdbc:postgresql://server/database", "jdbc:postgresql://server:5432/database");
    testFailoverUrl("jdbc:postgresql://server1,server2/database", "jdbc:postgresql://server1:5432,server2:5432/database");
    // ... add test cases here ...
}
```

Reply:
But then the test will always run for `BaseDataSource`, not for the concrete subclass (like `PoolingDataSource`) under test in the `BaseDataSourceTest` test subclass.

Maybe it is sufficient to set dbs to null after the test, as the other tests either call `initializeDataSource()` to set up a new one anyway, or `getDataSourceConnection()` which will do the same if `bds` is null ...?

Reply:
ok setting null was not a good idea, so how about preserving the original instance ? (see latest commit)

Reply:
@sehrope the BaseDataSource is abstract, we can't do `new BaseDataSource();` as you suggested.
But I'm also not getting anywhere with my ideas as the `PoolingDataSourceTest` does not allow a second instance.
Any suggestions how we could test the concrete BaseDataSource subclass under test when we need to modify it's parameters ?
Or I will just test the `PGSimpleDataSource' as it's the simplest concrete implementation ?